### DISCUSSION TITLE
Too much runtime for a DFN + Thermal model simulation

### DISCUSSION THREAD

**Discussion by abhishek-appana** [pybamm-team/PyBaMM#4296] [Q&A] [answered]
Hello everyone,

Can anyone give me an idea of how long it will take to run DFN + Thermal model simulation? Here are more details of the setup! I left the simulation running for a day and it's still going on. I noticed on this [link](https://pybamm-team.github.io/pybamm-bench/#summarylist?sort=0&dir=asc) that the airspeed velocity is very low for this simulation. I don't understand why it is taking so much time for me!

Chemistry:  NCA_Kim2011
Thermal model: Pouch cell - 2D current collectors [Options: "Dimensionality": 2, "Thermal": x-lumped]

I am running this simulation on a desktop with the following specs:
RAM: 32 GB
CPU Base speed: 2.7 GHz
L1 Cache: 1.2 MB
GPU memory: 16 GB

Below is the minimal working example:

```python
import pybamm
import os
import pandas as pd
from pybamm.expression_tree.functions import exp
os.chdir(pybamm.__path__[0]+'/..')
import numpy as np
import timeit
from scipy import interpolate


experiment = pybamm.Experiment(
    [
        (
            "Discharge at C/20 for 24 hours or until 3 V",
        )
    ]
)

options = {"dimensionality": 2, "thermal": "x-lumped"}
model = pybamm.lithium_ion.DFN(options=options)
parameter_values = pybamm.ParameterValues("NCA_Kim2011")

solver = pybamm.CasadiSolver(mode="safe",dt_max=10)
sim = pybamm.Simulation(model,parameter_values=parameter_values, solver=solver,experiment=experiment)
sim.solve()
```

**Comment by MarcBerliner** [comment, marked as answer]
Hi @abhishek-appana, I tried running the code but also did not get a solution within 15 minutes or so.

We are actively working on improving the solvers in PyBaMM. Can you try using the `develop` branch of PyBaMM and switching to `IDAKLUSolver`, i.e.
```python
solver = pybamm.IDAKLUSolver()
sim = pybamm.Simulation(model,parameter_values=parameter_values, solver=solver,experiment=experiment)
sim.solve()
```
With the new solver, I got an answer in 2 minutes and 15 seconds on my machine.

**Reply by abhishek-appana** [reply]
Hello @MarcBerliner ,

Thank you for your response! Understood!

I run a Windows system. I tried installing the IDAKLU solver earlier but was not able to. Do you have any resources that I can follow to do the installation?

**Reply by agriyakhetarpal** [reply]
Hi @abhishek-appana, installing the IDAKLU solver on Windows is tricky and we haven't added instructions to do so, yet – but we plan to do so with #4242 and other follow-ups. A short guide is as follows:

1. Set up a C++ compiler: we use a recent version of MSVC, which you can get using [Build Tools for Visual Studio Code 2022](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2022). Note that you won't need Visual Studio 2022 entirely, just "Desktop development with C++" should be enough.
2. Next, you'll need the `vcpkg` package manager for C++. You may get it here: https://vcpkg.io/en/. `vcpkg` should set up all of the prerequisite dependencies for you.
3. You can then do `pip install -ve .` inside a git clone of the PyBaMM repository. Before that, you'll need to set a bunch of environment variables as follows: https://github.com/pybamm-team/PyBaMM/blob/48b4e48d776ad2c254e5fae8db94706c57a8cc38/.github/workflows/publish_pypi.yml#L78-L83

In any case, the easier method, if you have elevated privileges on your machine, is to use WSL2: https://docs.pybamm.org/en/stable/source/user_guide/installation/windows-wsl.html and follow the instructions for GNU/Linux after that.

**Reply by abhishek-appana** [reply]
Thank you for the guide @agriyakhetarpal!

Will try it and reach out if I run into issues!

**Reply by abhishek-appana** [reply]
Hello @agriyakhetarpal ,

I followed the easier method you suggested (using WSL2). When I run the tests using  ``` nox ``` , I get the following error. How do I resolve this?


E               RuntimeError: latex program is not installed

.nox/unit/lib/python3.10/site-packages/sympy/printing/preview.py:308: RuntimeError
============================================================================================ short test summary info ============================================================================================
SKIPPED [1] tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_newman_tobias.py:19: Test currently not implemented
SKIPPED [1] tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_newman_tobias.py:23: Test currently not implemented
SKIPPED [1] tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_newman_tobias.py:27: Test currently not implemented
SKIPPED [1] tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_newman_tobias.py:31: Test currently not implemented
SKIPPED [1] tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_newman_tobias.py:35: Test currently not implemented
SKIPPED [1] tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_newman_tobias.py:39: Test currently not implemented
SKIPPED [1] tests/unit/test_solvers/test_idaklu_jax.py:91: Both IDAKLU and JAX are available
SKIPPED [1] tests/unit/test_util.py:91: The JAX solver is not installed
FAILED tests/unit/test_expression_tree/test_operations/test_latexify.py::TestLatexify::test_sympy_preview - RuntimeError: latex program is not installed
============================================================================= 1 failed, 1653 passed, 8 skipped in 188.06s (0:03:08) =============================================================================
An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
nox > Command python -m pytest -m unit failed with exit code 1
nox > Session unit failed.

**Reply by agriyakhetarpal** [reply]
Hi @abhishek-appana, this error means that you don't have a LaTeX preprocessor installed. Since the rest of the tests are passing, you can safely ignore this – PyBaMM has been installed correctly, and the IDAKLU solver should now be usable. The only functionality for which LaTeX is required is for the `model.latexify()` API that allows printing the equations for the model it is called for.

**Reply by abhishek-appana** [reply]
Thank you for the response @agriyakhetarpal .

Thought the same but wanted to confirm! I am able to use IDAKLU solver and run simulations. Thank you so much for the help!

**Reply by abhishek-appana** [reply]
Hello @agriyakhetarpal ,

I just noticed that if I intend to make edits to the source code and use those edits, I need to install PyBaMM in editable mode.

When I run  ``` pip install -e .[all,dev,docs] ```  in the PyBaMM directory, it doesn't get installed properly. Can you help me with this?

Am I supposed to do the manual install inside the virtual environment which is created when I run ``` nox -s dev ``` ?

____________________________________________________________________________________

Obtaining file:///root/PyBaMM
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
WARNING: unknown 0.0.0 does not provide the extra 'all'
WARNING: unknown 0.0.0 does not provide the extra 'dev'
WARNING: unknown 0.0.0 does not provide the extra 'docs'
Installing collected packages: UNKNOWN
  Attempting uninstall: UNKNOWN
    Found existing installation: UNKNOWN 0.0.0
    Uninstalling UNKNOWN-0.0.0:
      Successfully uninstalled UNKNOWN-0.0.0
  Running setup.py develop for UNKNOWN
Successfully installed UNKNOWN-0.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv

**Reply by agriyakhetarpal** [reply]
Hi @abhishek-appana, this looks strange. I tried with the latest changes from `develop`, and I don't seem to have this error – could you please try again in a fresh virtual environment, and if the error doesn't subside, open a new issue with the steps you followed?

**Reply by abhishek-appana** [reply]
I am doing the pip install in the PyBaMM directory locally. I am not doing it inside the virtual environment. Am I supposed to do it inside a virtual environment?

**Reply by agriyakhetarpal** [reply]
A virtual environment would be good, yes. FTR, after you do `nox -s pybamm-requires`, you can do `nox -s dev`, which sets up a virtual environment for you in the `venv/` directory, which you can simply activate using `source venv/bin/activate`.

**Reply by abhishek-appana** [reply]
I am able to install it inside the virtual environment. Thank you!

The installation guide on the [website](https://docs.pybamm.org/en/stable/source/user_guide/installation/install-from-source.html#installing-pybamm) doesn't specify that! That's why I was doing in the PyBaMM directory

Thanks for your help!

**Reply by agriyakhetarpal** [reply]
Thanks, our installation docs definitely need aome improvements. I see that we do specify virtual environments here: https://docs.pybamm.org/en/stable/source/user_guide/installation/windows.html#user-install buf not for the source installation instructions.

**Reply by abhishek-appana** [reply]
I see. Thank you 😀

**Reply by benshang1997** [reply]
Hi @abhishek-appana ,  thanks for your help, i will try.

**Reply by benshang1997** [reply]
> Hi @abhishek-appana, this error means that you don't have a LaTeX preprocessor installed. Since the rest of the tests are passing, you can safely ignore this – PyBaMM has been installed correctly, and the IDAKLU solver should now be usable. The only functionality for which LaTeX is required is for the API that allows printing the equations for the model it is called for.`model.latexify()`
<img width="857" alt="1723966688512" src="https://github.com/user-attachments/assets/654227b6-38a1-443d-9db0-26ca733597a8">

Hi @agriyakhetarpal thanks for your suggestion, I think I meet the same problem, Do I install it successfully too? and how should I do if i still want to install the latex? thankyou

And I can't find run-tests.py to run the test.  but I have a test folder with many sample。 so I wonder if the run-test has been changed name?
<img width="729" alt="1723969696854" src="https://github.com/user-attachments/assets/bc9e1557-6995-47d6-9360-e9ae5bcf8888">

**Reply by agriyakhetarpal** [reply]
Hi @benshang1997, you can skip this test safely. However, if you still wish to install a LaTeX distribution, you can try doing `sudo apt-get install texlive-full` (you might not need the full distribution, so you can replace it with the one you need, or a smaller one).

We remove the `run-tests.py` file, so the tests can be triggered entirely using `pytest` (and `nox`). Here's the relevant section in the Contributing guide: https://docs.pybamm.org/en/latest/source/user_guide/contributing.html#testing (note that this is from the `latest` section of the docs because it hasn't been released in the `stable` version yet).

**Reply by benshang1997** [reply]
> 嗨，您可以安全地跳过此测试。但是，如果您仍然希望安装 LaTeX 发行版，您可以尝试这样做（您可能不需要完整的发行版，因此您可以用您需要的发行版或更小的发行版替换它）。`sudo apt-get install texlive-full`
> 
> 我们删除了该文件，因此可以完全使用 （和 ） 触发测试。以下是贡献指南中的相关部分： https://docs.pybamm.org/en/latest/source/user_guide/contributing.html#testing（请注意，这是来自文档的部分，因为它尚未在版本中发布）。`run-tests.py``pytest``nox``latest``stable`

Hi @ @agriyakhetarpal  thanks for your guide I have solved this issued, but the solver maybe meet a the convergence issues as belowed when I run expriement:

At t = 279.719 and h = 1.39219e-11, the corrector convergence failed repeatedly or with |h| = hmin.

but the experiment is not stopped with this issue. so I wonder does this impact the simulation result? and could you please give me some advices on it,  i also met the same issues in casadi and that is why i change to use IDA

**Reply by MarcBerliner** [reply]
> At t = 279.719 and h = 1.39219e-11, the corrector convergence failed repeatedly or with |h| = hmin.

Hi @benshang1997, I'm just double-checking. Are you using the `develop` branch of PyBaMM? This works on my machine using the `develop` branch (not the latest PyBaMM release).

**Comment by benshang1997** [comment]
Hi @abhishek-appana I am install a new env and try to install IDAKLU according the guide https://docs.pybamm.org/en/stable/source/user_guide/installation/install-from-source.html.

But I meet some issues, I can clone the GitHub repository successfully but I cant run this :sudo apt install python3.X python3.X-dev libopenblas-dev gcc gfortran graphviz cmake pandoc
<img width="874" alt="1723935446265" src="https://github.com/user-attachments/assets/e4e60d37-bf0a-4852-afef-cb2e7d9c205d">

I cant use nox if I skip this.
(pybamm_ubun is my new env and it is empty before I clone pybamm)

could you please give me some suggestion? I am a starter on linux too. 
thanks a million

**Reply by abhishek-appana** [reply]
Hi @benshang1997 ,

I don't have experience with linux either. Are you using wsl?

Try running `sudo apt update && upgrade` before running the install command. I don't know if it'll help but just a try! I don't have any idea as to why this is happening

**Reply by agriyakhetarpal** [reply]
You might not receive updated versions of packages such as `python3.12-dev` from `apt`. I would suggest installing via a different package manager, say, something like Linuxbrew or getting it from the official releases on https://www.python.org/downloads/. Another method would be to use conda/mamba, which is probably the easiest method.
